File: mapeable/Mapeable/Comparador/ComparadorBaseDeColecciones.vb

Class ComparadorBaseDeColecciones

    Dim sonIguales As Boolean
    Dim estaColeccion As IEnumerable(Of Object)
    Dim laOtraColeccion As IEnumerable(Of Object)

    Function EsIgualQueLaColeccion(estaColeccion As IEnumerable(Of Object),
                                   laOtraColeccion As IEnumerable(Of Object)) As Boolean
        Me.estaColeccion = estaColeccion
        Me.laOtraColeccion = laOtraColeccion

        CompareSiSonNulos()
        CompareLasColecciones()

        Return sonIguales
    End Function

    Private Sub CompareSiSonNulos()
        If estaColeccion Is Nothing And laOtraColeccion Is Nothing Then
            sonIguales = True
        End If
    End Sub

    Private Sub CompareLasColecciones()
        If NoHayNulos() AndAlso SusTiposSonIguales() AndAlso SusTamanosSonIguales() Then

            sonIguales = True
            For i As Integer = 0 To estaColeccion.Count - 1

                Dim unObjecto As Object
                Dim otroObjecto As Object

                unObjecto = estaColeccion.ElementAt(i)
                otroObjecto = laOtraColeccion.ElementAt(i)

                Dim elComparador As New ComparadorBase()
                sonIguales = sonIguales And elComparador.EsIgualQue(unObjecto, otroObjecto)
            Next
        End If
    End Sub

    Private Function NoHayNulos() As Boolean
        Dim hayNulos As Boolean
        hayNulos = estaColeccion Is Nothing Or laOtraColeccion Is Nothing
        Return Not hayNulos
    End Function

    Private Function SusTiposSonIguales() As Boolean
        Dim sonElMismoTipo As Boolean
        If NoHayNulos() Then
            Dim elTipodeEstaColeccion As Type
            elTipodeEstaColeccion = estaColeccion.GetType
            Dim elTipoDeLaOtraColeccion As Type
            elTipoDeLaOtraColeccion = laOtraColeccion.GetType
            sonElMismoTipo = Type.Equals(elTipodeEstaColeccion, elTipoDeLaOtraColeccion)
        End If
        Return sonElMismoTipo
    End Function

    Private Function SusTamanosSonIguales()
        Dim tienenElMismoTamano As Boolean
        If NoHayNulos() And SusTiposSonIguales() Then
            tienenElMismoTamano = DetermineSiTienenElMismoTamano()
        End If
        Return tienenElMismoTamano
    End Function

    Private Function DetermineSiTienenElMismoTamano() As Boolean
        Dim tienenElMismoTamano As Boolean
        If estaColeccion.Count = laOtraColeccion.Count Then
            tienenElMismoTamano = True
        Else
            tienenElMismoTamano = False
        End If
        Return tienenElMismoTamano
    End Function

End Class

File: mapeable/Mapeable.ComparacionesParaPruebasUnitarias/Comparaciones/ComparadorBaseDeColecciones.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace Comparaciones
{
    public class ComparadorBaseDeColecciones
    {
        private bool sonIguales;
        private IEnumerable<object> estaColeccion;

        private IEnumerable<object> laOtraColeccion;
        public bool EsIgualQueLaColeccion(IEnumerable<object> estaColeccion, 
            IEnumerable<object> laOtraColeccion)
        {
            this.estaColeccion = estaColeccion;
            this.laOtraColeccion = laOtraColeccion;

            CompareSiSonNulos();
            CompareLasColecciones();

            return sonIguales;
        }

        private void CompareSiSonNulos()
        {
            if (estaColeccion == null & laOtraColeccion == null)
            {
                sonIguales = true;
            }
        }

        private void CompareLasColecciones()
        {
            if (NoHayNulos() && SusTiposSonIguales() && SusTamanosSonIguales())
            {
                sonIguales = true;

                for (int i = 0; i <= estaColeccion.Count() - 1; i++)
                {
                    object unObjecto = null;
                    object otroObjecto = null;

                    unObjecto = estaColeccion.ElementAt(i);
                    otroObjecto = laOtraColeccion.ElementAt(i);

                    ComparadorBase elComparador = new ComparadorBase();
                    sonIguales = sonIguales & elComparador.EsIgualQue(unObjecto, otroObjecto);
                }
            }
        }

        private bool NoHayNulos()
        {
            bool hayNulos = false;
            hayNulos = estaColeccion == null | laOtraColeccion == null;
            return !hayNulos;
        }

        private bool SusTiposSonIguales()
        {
            bool sonElMismoTipo = false;
            if (NoHayNulos())
            {
                Type elTipodeEstaColeccion = estaColeccion.GetType();
                Type elTipoDeLaOtraColeccion = laOtraColeccion.GetType();
                sonElMismoTipo = Type.Equals(elTipodeEstaColeccion, elTipoDeLaOtraColeccion);
            }
            return sonElMismoTipo;
        }

        private bool SusTamanosSonIguales()
        {
            bool tienenElMismoTamano = false;
            if (NoHayNulos() & SusTiposSonIguales())
            {
                tienenElMismoTamano = DetermineSiTienenElMismoTamano();
            }
            return tienenElMismoTamano;
        }

        private bool DetermineSiTienenElMismoTamano()
        {
            bool tienenElMismoTamano = false;
            if (estaColeccion.Count() == laOtraColeccion.Count())
            {
                tienenElMismoTamano = true;
            }
            else {
                tienenElMismoTamano = false;
            }
            return tienenElMismoTamano;
        }

    }
}